<a href="https://colab.research.google.com/github/Valerie-Osawe/predict_salary/blob/main/Assignment_7.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Predict A Person's Salary



### Data Dictionary

```
- Age: continuous.

- Workclass: Private, Self-emp-not-inc, Self-emp-inc, Federal-gov, Local-gov, State-gov, Without-pay, Never-worked.

- Final_weight: continuous.

- Education: Bachelors, Some-college, 11th, HS-grad, Prof-school, Assoc-acdm, Assoc-voc, 9th, 7th-8th, 12th, - Masters, 1st-4th, 10th, Doctorate, 5th-6th, Preschool.

- Education_num: continuous.

- Marital_status: Married-civ-spouse, Divorced, Never-married, Separated, Widowed, Married-spouse-absent, - Married-AF-spouse.

- Occupation: Tech-support, Craft-repair, Other-service, Sales, Exec-managerial, Prof-specialty, Handlers-cleaners, Machine-op-inspct, Adm-clerical, Farming-fishing, Transport-moving, Priv-house-serv, Protective-serv, Armed-Forces.

- Relationship: Wife, Own-child, Husband, Not-in-family, Other-relative, Unmarried.

- Race: White, Asian-Pac-Islander, Amer-Indian-Eskimo, Other, Black.

- Genger: Female, Male.

- Capital_gain: continuous.

- Capital_loss: continuous.

- Hours_per_week: continuous.

- Country: United-States, Cambodia, England, Puerto-Rico, Canada, Germany, Outlying-US(Guam-USVI-etc), India, Japan, Greece, South, China, Cuba, Iran, Honduras, Philippines, Italy, Poland, Jamaica, Vietnam, Mexico, Portugal, Ireland, France, Dominican-Republic, Laos, Ecuador, Taiwan, Haiti, Columbia, Hungary, Guatemala, Nicaragua, Scotland, Thailand, Yugoslavia, El-Salvador, Trinadad&Tobago, Peru, Hong, Holand-Netherlands.

- Salary: 
```

### Objective

```
Predict whether a person makes over 50K a year.

```

In [52]:
# Built-in library
import itertools

# Standard imports
import numpy as np
import pandas as pd


# pandas settings
pd.options.display.max_rows = 1_000
pd.options.display.max_columns = 1_000
pd.options.display.max_colwidth = 1_000



In [53]:
# Load the dataset
salary = pd.read_csv("salary.csv")

# Check the first few rows of the dataset
salary.head()

Age          Workclass  Final_weight   Education  Education_num  \
0   50   Self-emp-not-inc         83311   Bachelors             13   
1   38            Private        215646     HS-grad              9   
2   53            Private        234721        11th              7   
3   28            Private        338409   Bachelors             13   
4   37            Private        284582     Masters             14   

        Marital_status          Occupation    Relationship    Race      Sex  \
0   Married-civ-spouse     Exec-managerial         Husband   White     Male   
1             Divorced   Handlers-cleaners   Not-in-family   White     Male   
2   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
3   Married-civ-spouse      Prof-specialty            Wife   Black   Female   
4   Married-civ-spouse     Exec-managerial            Wife   White   Female   

   Capital_gain  Capital_loss  Hours_per_week         Country  Salary  
0             0             0              13   United-States   <=50K  
1             0             0              40   United-States   <=50K  
2             0             0              40   United-States   <=50K  
3             0             0              40            Cuba   <=50K  
4             0             0              40   United-States   <=50K

### Data cleaning and Preprocessing

 The data needs to cleaned and preprocessed the before we can use it to train our machine learning model. This will involve handling missing values, converting categorical variables to numerical.



In [54]:
# Check the shape of the dataset
salary.shape

(32560, 15)

In [55]:
# Check for missing values
salary.isnull().sum()

Age               0
Workclass         0
Final_weight      0
Education         0
Education_num     0
Marital_status    0
Occupation        0
Relationship      0
Race              0
Sex               0
Capital_gain      0
Capital_loss      0
Hours_per_week    0
Country           0
Salary            0
dtype: int64

In [56]:
# explore the country column
salary.Country.value_counts()

 United-States                 29169
 Mexico                          643
 ?                               583
 Philippines                     198
 Germany                         137
 Canada                          121
 Puerto-Rico                     114
 El-Salvador                     106
 India                           100
 Cuba                             95
 England                          90
 Jamaica                          81
 South                            80
 China                            75
 Italy                            73
 Dominican-Republic               70
 Vietnam                          67
 Guatemala                        64
 Japan                            62
 Poland                           60
 Columbia                         59
 Taiwan                           51
 Haiti                            44
 Iran                             43
 Portugal                         37
 Nicaragua                        34
 Peru                             31
 

In [57]:
# Dropping rows with '?' in the dataframe

# Replace '?' with NaN
salary.replace(' ?', np.nan, inplace=True)

# Drop rows with NaN values
salary.dropna(inplace=True)

In [58]:
# Check the distribution of the target variable
salary["Salary"].value_counts()

 <=50K    22653
 >50K      7508
Name: Salary, dtype: int64

In [59]:
from sklearn.preprocessing import LabelEncoder

cat_vars = [
    "Workclass",
    "Education",
    "Marital_status",
    "Occupation",
    "Relationship",
    "Race",
    "Sex",
    "Country",
]
# Convert categorical variables to dummy variables
salary = pd.get_dummies(salary, columns=cat_vars)


salary.head()

Age  Final_weight  Education_num  Capital_gain  Capital_loss  \
0   50         83311             13             0             0   
1   38        215646              9             0             0   
2   53        234721              7             0             0   
3   28        338409             13             0             0   
4   37        284582             14             0             0   

   Hours_per_week  Salary  Workclass_ Federal-gov  Workclass_ Local-gov  \
0              13   <=50K                       0                     0   
1              40   <=50K                       0                     0   
2              40   <=50K                       0                     0   
3              40   <=50K                       0                     0   
4              40   <=50K                       0                     0   

   Workclass_ Private  Workclass_ Self-emp-inc  Workclass_ Self-emp-not-inc  \
0                   0                        0                            1   
1                   1                        0                            0   
2                   1                        0                            0   
3                   1                        0                            0   
4                   1                        0                            0   

   Workclass_ State-gov  Workclass_ Without-pay  Education_ 10th  \
0                     0                       0                0   
1                     0                       0                0   
2                     0                       0                0   
3                     0                       0                0   
4                     0                       0                0   

   Education_ 11th  Education_ 12th  Education_ 1st-4th  Education_ 5th-6th  \
0                0                0                   0                   0   
1                0                0                   0                   0   
2                1                0                   0                   0   
3                0                0                   0                   0   
4                0                0                   0                   0   

   Education_ 7th-8th  Education_ 9th  Education_ Assoc-acdm  \
0                   0               0                      0   
1                   0               0                      0   
2                   0               0                      0   
3                   0               0                      0   
4                   0               0                      0   

   Education_ Assoc-voc  Education_ Bachelors  Education_ Doctorate  \
0                     0                     1                     0   
1                     0                     0                     0   
2                     0                     0                     0   
3                     0                     1                     0   
4                     0                     0                     0   

   Education_ HS-grad  Education_ Masters  Education_ Preschool  \
0                   0                   0                     0   
1                   1                   0                     0   
2                   0                   0                     0   
3                   0                   0                     0   
4                   0                   1                     0   

   Education_ Prof-school  Education_ Some-college  Marital_status_ Divorced  \
0                       0                        0                         0   
1                       0                        0                         1   
2                       0                        0                         0   
3                       0                        0                         0   
4                       0                        0                         0   

   Marital_status_ Married-AF-spouse  Marital_status_ Married-civ-spouse  \
0                 

In [60]:
# Scale the numerical variables
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
salary[
    [
        "Age",
        "Final_weight",
        "Education_num",
        "Capital_gain",
        "Capital_loss",
        "Hours_per_week",
    ]
] = scaler.fit_transform(
    salary[
        [
            "Age",
            "Final_weight",
            "Education_num",
            "Capital_gain",
            "Capital_loss",
            "Hours_per_week",
        ]
    ]
)
salary.head()

Age  Final_weight  Education_num  Capital_gain  Capital_loss  \
0  0.880275     -1.007909       1.128961     -0.147437      -0.21859   
1 -0.033338      0.244659      -0.439703     -0.147437      -0.21859   
2  1.108678      0.425206      -1.224035     -0.147437      -0.21859   
3 -0.794682      1.406626       1.128961     -0.147437      -0.21859   
4 -0.109472      0.897147       1.521127     -0.147437      -0.21859   

   Hours_per_week  Salary  Workclass_ Federal-gov  Workclass_ Local-gov  \
0       -2.331495   <=50K                       0                     0   
1       -0.077735   <=50K                       0                     0   
2       -0.077735   <=50K                       0                     0   
3       -0.077735   <=50K                       0                     0   
4       -0.077735   <=50K                       0                     0   

   Workclass_ Private  Workclass_ Self-emp-inc  Workclass_ Self-emp-not-inc  \
0                   0                        0                            1   
1                   1                        0                            0   
2                   1                        0                            0   
3                   1                        0                            0   
4                   1                        0                            0   

   Workclass_ State-gov  Workclass_ Without-pay  Education_ 10th  \
0                     0                       0                0   
1                     0                       0                0   
2                     0                       0                0   
3                     0                       0                0   
4                     0                       0                0   

   Education_ 11th  Education_ 12th  Education_ 1st-4th  Education_ 5th-6th  \
0                0                0                   0                   0   
1                0                0                   0                   0   
2                1                0                   0                   0   
3                0                0                   0                   0   
4                0                0                   0                   0   

   Education_ 7th-8th  Education_ 9th  Education_ Assoc-acdm  \
0                   0               0                      0   
1                   0               0                      0   
2                   0               0                      0   
3                   0               0                      0   
4                   0               0                      0   

   Education_ Assoc-voc  Education_ Bachelors  Education_ Doctorate  \
0                     0                     1                     0   
1                     0                     0                     0   
2                     0                     0                     0   
3                     0                     1                     0   
4                     0                     0                     0   

   Education_ HS-grad  Education_ Masters  Education_ Preschool  \
0                   0                   0                     0   
1                   1                   0                     0   
2                   0                   0                     0   
3                   0                   0                     0   
4                   0                   1                     0   

   Education_ Prof-school  Education_ Some-college  Marital_status_ Divorced  \
0                       0                        0                         0   
1                       0                        0                         1   
2                       0                        0                         0   
3                       0                        0                         0   
4                       0                        0                         0   

   Marital_status_ Married-AF-spouse  Marital_status_ Married-civ-spo

The columns originally with numeric values were standardized. This is because Standardization of data improves performance of some models  can help to reduce the impact of skewed distributions or extreme values.

### Train model

Now that the data has been preprocessed, a machine learning model can be trained to predict a person's salary based on the other features in the dataset. To do this, logistic regression model will be used, this is because the target variable, salary, is binary (whether a person earns more than 50K or not) and logistic regression is a suitable algorithm for binary classification problems. It estimates the probability of the target variable based on the input features, and then applies a threshold to make a binary classification decision. Logistic regression is also a simple and interpretable model, and can handle both numerical and categorical features.



.

In [61]:
from sklearn.model_selection import train_test_split

# Split the data into features (X) and target (y)
X = salary.drop(columns=["Salary"])
y = salary["Salary"]

# Split the data into a training set and a test set
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

 20% of the data is used for testing while 80% for training while ensuring the same split can be obtained if the code is run again with the same random seed (42 in this case).

In [62]:
from sklearn.linear_model import LogisticRegression

# Train a logistic regression model
model = LogisticRegression()
model.fit(X_train, y_train)

/usr/local/lib/python3.9/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

### Evaluate Model

The model was evaluated to determine how well it performs on predicting a person's salary and assess the quality of the model's predictions. 

In [63]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Make predictions on the test set
predict = model.predict(X_test)

# Calculate the accuracy
print("Accuracy: ", accuracy_score(y_test, predict))

Accuracy:  0.8458478368970661


In [64]:
# Calculate the precision
print("Precision: ", precision_score(y_test, predict, pos_label=" >50K"))

Precision:  0.7290322580645161


In [65]:
# Calculate the recall score
print("Recall: ", recall_score(y_test, predict, pos_label=" >50K"))

# Calculate the f1 score
print("F1-Score: ", f1_score(y_test, predict, pos_label=" >50K"))

Recall:  0.6034712950600801
F1-Score:  0.660336011687363


The accuracy, precision, recall, and F1-score values of the model indicates that the model has performed moderately well. The accuracy of the model is 84.58%, which means that the model is able to predict the correct class label for 84.58% of the instances in the test dataset.

The precision of the model is 72.9%, which means that out of all the instances that the model has predicted as positive, 72.9% of them are actually positive.

The recall of the model is 60.35%, which means that out of all the actual positive instances in the test dataset, the model is able to correctly identify 60.35% of them.

The F1-score of the model is 66.03%, which is the harmonic mean of precision and recall. It is a balanced measure that takes both precision and recall into account.

Based on the metrics, the model is doing reasonably well but there is still room for improvement.

### Predict Salary

In [66]:
# Create a new candidate
new_candidate = pd.DataFrame(
    {
        "Age": [30],
        "Workclass": ["Private"],
        "Final_weigh": [77516],
        "Education": ["Bachelors"],
        "Education_num": [13],
        "Marital_status": ["Never-married"],
        "Occupation": ["Adm-clerical"],
        "Relationship": ["Not-in-family"],
        "Race": ["White"],
        "Sex": ["Male"],
        "Capital-gain": [0],
        "Capital-loss": [0],
        "Hours-per-week": [40],
        "Country": ["United-States"],
    }
)

# Convert categorical variables to dummy variables
new_candidate = pd.get_dummies(new_candidate, columns=cat_vars)

# Ensure that the new_candidate columns are in the same order as the training data
new_candidate = new_candidate.reindex(columns=X_train.columns, fill_value=0)

# Predict the salary of a person
new_salary = model.predict(new_candidate)

if new_salary == " >50K":
    print("The person makes over 50K a year")
else:
    print("The person doesn't make over 50K a year")

The person makes over 50K a year


The above code is an attempt to predict the salary of a new candidate based on a logistic regression model trained on the data.

The code first creates a pandas DataFrame named new_candidate containing the values of the features of the new candidate, including Age, Workclass, Final_weigh, Education, Education_num, Marital_status, Occupation, Relationship, Race, Sex, Capital-gain, Capital-loss, Hours-per-week, and Country.

Then, the code attempts to predict the salary of the new candidate using the predict method of the logistic regression model that was previously trained. However, since the feature names in new_candidate do not match the feature names used during training, the code raises a ValueError.

To fix this, the feature names in new_candidate should be changed to match the feature names used during training. To do this, the values for the categorical features should be one-hot encoded, similar to how it was done during training.